harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -67,110 +68,114 @@ def test_require_global_batch_size(self) -> None:
     def test_onevar_single(self) -> None:
         """Assert that the training loss and validation error decrease monotonically."""
 
-        def make_workloads() -> workload.Stream:
-            trainer = utils.TrainAndValidate()
+        class TrainingMetricsCallback(pytorch.PyTorchCallback):
+            def __init__(self):
+                self.steps = 0
+                self.metrics = []
 
-            yield from trainer.send(steps=100, validation_freq=10)
-            training_metrics, validation_metrics = trainer.result()
-
-            # Check the gradient update at every step.
-            for idx, batch_metrics in enumerate(training_metrics):
+            def on_training_workload_end(
+                self, avg_metrics: typing.Dict[str, typing.Any], batch_metrics: typing.Dict[str, typing.Any]
+            ) -> None:
+                self.metrics.append(batch_metrics[0])
                 pytorch_onevar_model.OneVarTrial.check_batch_metrics(
-                    batch_metrics,
-                    idx,
+                    batch_metrics[0],
+                    self.steps,
                     metric_keyname_pairs=(("loss", "loss_exp"), ("w_after", "w_exp")),
                 )
+                self.steps += 1
 
-            for older, newer in zip(training_metrics, training_metrics[1:]):
-                assert newer["loss"] <= older["loss"]
+            def on_trial_shutdown(self) -> None:
+                for older, newer in zip(self.metrics, self.metrics[1:]):
+                    assert newer["loss"] <= older["loss"]
+
+        def build_callbacks(self) -> typing.Dict[str, pytorch.PyTorchCallback]:
+            return {"metricsCallback": TrainingMetricsCallback()}
+
+        setattr(pytorch_onevar_model.OneVarTrial, "build_callbacks", build_callbacks)
```

Comment:
should callbacks be passed in when you call `train` instead of being set on the Trial?

Reply:
I'm fine landing it the way it is.

Some day we might consider migrating to passing them in `fit()`, or making it possible to pass in `fit()` or Trial definition.  But we don't have to decide today.